x-pack/plugins/security/server/routes/authorization/roles/put.test.ts

```diff
@@ -0,0 +1,852 @@
+
+import {
+  elasticsearchServiceMock,
+  httpServerMock,
+} from '../../../../../../../src/core/server/mocks';
+import { routeDefinitionParamsMock } from '../../index.mock';
+
+const application = 'kibana-.kibana';
+const privilegeMap = {
+  global: {
+    all: [],
+    read: [],
+  },
+  space: {
+    all: [],
+    read: [],
+  },
+  features: {
+    foo: {
+      'foo-privilege-1': [],
+      'foo-privilege-2': [],
+    },
+    bar: {
+      'bar-privilege-1': [],
+      'bar-privilege-2': [],
+    },
+  },
+  reserved: {
+    customApplication1: [],
+    customApplication2: [],
+  },
+};
+
+interface TestOptions {
+  name: string;
+  licenseCheckResult?: ILicenseCheck;
+  apiResponses?: Array<() => Promise<unknown>>;
+  payload?: Record<string, any>;
+  asserts: { statusCode: 204 | 403; result?: Record<string, any>; apiArguments?: unknown[][] };
+}
+
+const putRoleTest = (
+  description: string,
+  {
+    name,
+    payload,
+    licenseCheckResult = { check: LICENSE_STATUS.Valid },
+    apiResponses = [],
+    asserts,
+  }: TestOptions
+) => {
+  test(description, async () => {
+    const mockRouteDefinitionParams = routeDefinitionParamsMock.create();
+    mockRouteDefinitionParams.authz.getApplicationName.mockReturnValue(application);
+    mockRouteDefinitionParams.authz.privileges.get.mockReturnValue(privilegeMap);
+
+    const mockScopedClusterClient = elasticsearchServiceMock.createScopedClusterClient();
+    mockRouteDefinitionParams.clusterClient.asScoped.mockReturnValue(mockScopedClusterClient);
+    for (const apiResponse of apiResponses) {
+      mockScopedClusterClient.callAsCurrentUser.mockImplementationOnce(apiResponse);
+    }
+
+    definePutRolesRoutes(mockRouteDefinitionParams);
+    const [[{ validate }, handler]] = mockRouteDefinitionParams.router.put.mock.calls;
+
+    const headers = { authorization: 'foo' };
+    const mockRequest = httpServerMock.createKibanaRequest({
+      method: 'put',
+      path: `/api/security/role/${name}`,
+      params: { name },
+      body: payload !== undefined ? (validate as any).body.validate(payload) : undefined,
+      headers,
+    });
+    const mockContext = ({
+      licensing: { license: { check: jest.fn().mockReturnValue(licenseCheckResult) } },
+    } as unknown) as RequestHandlerContext;
+
+    const mockResponseResult = { status: asserts.statusCode, options: {} };
+    const mockResponse = httpServerMock.createResponseFactory();
+    let mockResponseFactory;
+    if (asserts.statusCode === 204) {
+      mockResponseFactory = mockResponse.noContent;
+    } else if (asserts.statusCode === 403) {
+      mockResponseFactory = mockResponse.forbidden;
+    } else {
+      mockResponseFactory = mockResponse.customError;
+    }
+    mockResponseFactory.mockReturnValue(mockResponseResult);
+
+    const response = await handler(mockContext, mockRequest, mockResponse);
+
+    expect(response).toBe(mockResponseResult);
+    expect(mockResponseFactory).toHaveBeenCalledTimes(1);
```

Comment:
__note:__ while reviewing Larry's PR I realized I could just use `response.status` as core response mock creates real responses :man_facepalming: I'll update this altogether with any review related changes.

Reply:
>  as core response mock creates real responses

I was bit mistaken, response mocks don't create `KibanaResponse` instances. But core exports real `kibanaResponseFactory` so I switched to it.